src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,473 @@
+        private static bool ClosingDelimiterExistsIfNeeded(SyntaxNode currentNode, ref int lastDelimiterPosition)
+        {
+            switch (currentNode.Kind())
+            {
+                case SyntaxKind.ArgumentList:
+                    var argumentList = (ArgumentListSyntax)currentNode;
+                    if (argumentList.CloseParenToken.IsMissing)
+                    {
+                        return false;
+                    }
+                    else lastDelimiterPosition = argumentList.CloseParenToken.Span.End;
+                    return true;
+                case SyntaxKind.ParenthesizedExpression:
+                    var parenthesizedExpression = (ParenthesizedExpressionSyntax)currentNode;
+                    if (parenthesizedExpression.CloseParenToken.IsMissing)
+                    {
+                        return false;
+                    }
+                    else lastDelimiterPosition = parenthesizedExpression.CloseParenToken.Span.End;
+                    return true;
+                case SyntaxKind.BracketedArgumentList:
+                    var bracketedArgumentList = (BracketedArgumentListSyntax)currentNode;
+                    if (bracketedArgumentList.CloseBracketToken.IsMissing)
+                    {
+                        return false;
+                    }
+                    else lastDelimiterPosition = bracketedArgumentList.CloseBracketToken.Span.End;
+                    return true;
+                case SyntaxKind.ObjectInitializerExpression:
+                    var initializerExpressionSyntax = (InitializerExpressionSyntax)currentNode;
+                    if (initializerExpressionSyntax.CloseBraceToken.IsMissing)
+                    {
+                        return false;
+                    }
+                    else lastDelimiterPosition = initializerExpressionSyntax.CloseBraceToken.Span.End;
+                    return true;
+                default:
+                    // Type of node does not require a closing delimiter
+                    return true;
+            }
+        }
+
+        private bool ReachedSemicolonSyntax(SyntaxNode currentNode, ISyntaxFactsService syntaxFacts, ref bool finalDelimiterNeedsSemicolon)
+        {
+            finalDelimiterNeedsSemicolon = false;
+
+            if (currentNode.IsKind(SyntaxKind.DoStatement))
+            {
+                finalDelimiterNeedsSemicolon = true;
+                return true;
+            }
+
+            if (syntaxFacts.IsStatement(currentNode)
+                || currentNode.IsKind(SyntaxKind.GetAccessorDeclaration)
+                || currentNode.IsKind(SyntaxKind.SetAccessorDeclaration)
+                || currentNode.IsKind(SyntaxKind.LocalDeclarationStatement))
+            {
+                return true;
+            }
+
+            if (currentNode.IsKind(SyntaxKind.VariableDeclaration))
+            {
+                if (!currentNode.Ancestors().Any(n => n.IsKind(SyntaxKind.LocalDeclarationStatement)))
```

Comment:
❔  Blanket statements about `Ancestors()` tend to be problematic. Can this be replaced with a condition that checks specific ancestors?